src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -580,18 +577,51 @@ namespace stan {
         return;
       }
 
+       // check argument qualifiers
+      if (!decl.body_.is_no_op_statement()) {
+        function_signature_t decl_sig =
+          function_signatures::instance().get_definition(decl.name_, sig);
+        if (!decl_sig.first.is_ill_formed()) {
+          if (decl_sig.second.size() != arg_types.size()) {
+            error_msgs << "Declaration, definition mismatch for function "
+                       << decl.name_
+                       << ": declaration has " << arg_types.size()
+                       << " arguments, definition has " << arg_types.size()
+                       << "arguments." << std::endl;
+            pass = false;
+            return;
+          }
+          for (size_t i = 0; i < arg_types.size(); ++i) {
+            if (decl_sig.second[i].expr_type_ != arg_types[i].expr_type_
+                || decl_sig.second[i].data_only_ != arg_types[i].data_only_) {
+              error_msgs << "Declaration, definition mismatch for function "
+                         << decl.name_ << " argument " << (i + 1)
+                         << ": argument declared as "
+                         << (decl_sig.second[i].data_only_ ? "" : "data ")
+                         << decl_sig.second[i].expr_type_
+                         << ", defined as "
+                         << (arg_types[i].data_only_ ? "" : "data ")
```

Comment:
Don't need that space after "data" as you end with a period.

Reply:
good catch, thanks